{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3921"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>...</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>base-media</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>calcium-source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>...</td>\n",
       "      <td>42 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>...</td>\n",
       "      <td>42 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3815859</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>...</td>\n",
       "      <td>42 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4187550</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>...</td>\n",
       "      <td>42 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4400313</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→C</td>\n",
       "      <td>...</td>\n",
       "      <td>42 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  index  exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "0     9  42C    1    124        1         1       1.0    702352           DEL   \n",
       "1    10  42C    1    124        1         1       1.0   1308318           SNP   \n",
       "2    11  42C    1    124        1         1       1.0   3815859           DEL   \n",
       "3    12  42C    1    124        1         1       1.0   4187550           SNP   \n",
       "4    13  42C    1    124        1         1       1.0   4400313           SNP   \n",
       "\n",
       "  Sequence Change  ... temperature carbon-source                   supplement  \\\n",
       "0          Δ21 bp  ...  42 celsius    glucose(4)  NaCl(0.5g/L) trace elements   \n",
       "1             G→C  ...  42 celsius    glucose(4)  NaCl(0.5g/L) trace elements   \n",
       "2          Δ82 bp  ...  42 celsius    glucose(4)  NaCl(0.5g/L) trace elements   \n",
       "3             C→T  ...  42 celsius    glucose(4)  NaCl(0.5g/L) trace elements   \n",
       "4             A→C  ...  42 celsius    glucose(4)  NaCl(0.5g/L) trace elements   \n",
       "\n",
       "  strain-description taxonomy-id base-media  nitrogen-source  \\\n",
       "0                 WT      511145         M9         NH4Cl(1)   \n",
       "1                 WT      511145         M9         NH4Cl(1)   \n",
       "2                 WT      511145         M9         NH4Cl(1)   \n",
       "3                 WT      511145         M9         NH4Cl(1)   \n",
       "4                 WT      511145         M9         NH4Cl(1)   \n",
       "\n",
       "       phosphorous-source sulfur-source calcium-source  \n",
       "0  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  \n",
       "1  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  \n",
       "2  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  \n",
       "3  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  \n",
       "4  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  \n",
       "\n",
       "[5 rows x 48 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# The data scrapped from breseq reports can have different character encodings. Best just to normalize upfront.\n",
    "from unicodedata import normalize\n",
    "\n",
    "\n",
    "mut_df = pd.read_pickle('./data/4_10_with_uniq_midpts.pkl')\n",
    "\n",
    "mut_df[\"Sequence Change\"] = mut_df.apply(lambda m: normalize('NFKC', m[\"Sequence Change\"]), axis=1)\n",
    "mut_df[\"Details\"] = mut_df.apply(lambda m: normalize('NFKC', m[\"Details\"]), axis=1)\n",
    "\n",
    "display(len(mut_df), mut_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OBJECT_ID</th>\n",
       "      <th>OBJECT_SYNONYM_NAME</th>\n",
       "      <th>OS_INTERNAL_COMMENT</th>\n",
       "      <th>KEY_ID_ORG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>EG10001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>ECK4045</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>b4053</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>alr5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ECK120000002</td>\n",
       "      <td>b0764</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      OBJECT_ID OBJECT_SYNONYM_NAME  OS_INTERNAL_COMMENT KEY_ID_ORG\n",
       "0  ECK120000001             EG10001                  NaN      ECK12\n",
       "1  ECK120000001             ECK4045                  NaN      ECK12\n",
       "2  ECK120000001               b4053                  NaN      ECK12\n",
       "3  ECK120000001                alr5                  NaN      ECK12\n",
       "4  ECK120000002               b0764                  NaN      ECK12"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regdb_gene_df = pd.read_csv(\n",
    "    \"./data/regulondb10/gene.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "regdb_gene_df.columns = [\n",
    "    \"GENE_ID\",\n",
    "    \"GENE_NAME\",\n",
    "    \"GENE_POSLEFT\",\n",
    "    \"GENE_POSRIGHT\",\n",
    "    \"GENE_STRAND\",\n",
    "    \"GENE_SEQUENCE\",\n",
    "    \"GC_CONTENT\",\n",
    "    \"CRI_SCORE\",\n",
    "    \"GENE_NOTE\",\n",
    "    \"GENE_INTERNAL_COMMENT\",\n",
    "    \"KEY_ID_ORG\",\n",
    "    \"GENE_TYPE\"\n",
    "]\n",
    "\n",
    "\n",
    "gene_synonym_df = pd.read_csv(\n",
    "    \"./data/regulondb10/object_synonym.txt\",\n",
    "    sep=\"\\t\",\n",
    "    comment='#',\n",
    "    header=None,\n",
    "    quoting=3\n",
    ")\n",
    "gene_synonym_df.columns = [\"OBJECT_ID\", \"OBJECT_SYNONYM_NAME\", \"OS_INTERNAL_COMMENT\", \"KEY_ID_ORG\"]\n",
    "gene_synonym_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_df = mut_df.rename(columns={\"mutation target annotation\": \"Gene\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>...</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>base-media</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>calcium-source</th>\n",
       "      <th>clean genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>...</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>{nagA}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>...</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>{clsA}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3815859</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>...</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>{rph}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4187550</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>...</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>{rpoC}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4400313</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→C</td>\n",
       "      <td>...</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>{hfq}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 49 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  index  exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "0     9  42C    1    124        1         1       1.0    702352           DEL   \n",
       "1    10  42C    1    124        1         1       1.0   1308318           SNP   \n",
       "2    11  42C    1    124        1         1       1.0   3815859           DEL   \n",
       "3    12  42C    1    124        1         1       1.0   4187550           SNP   \n",
       "4    13  42C    1    124        1         1       1.0   4400313           SNP   \n",
       "\n",
       "  Sequence Change  ... carbon-source                   supplement  \\\n",
       "0          Δ21 bp  ...    glucose(4)  NaCl(0.5g/L) trace elements   \n",
       "1             G→C  ...    glucose(4)  NaCl(0.5g/L) trace elements   \n",
       "2          Δ82 bp  ...    glucose(4)  NaCl(0.5g/L) trace elements   \n",
       "3             C→T  ...    glucose(4)  NaCl(0.5g/L) trace elements   \n",
       "4             A→C  ...    glucose(4)  NaCl(0.5g/L) trace elements   \n",
       "\n",
       "  strain-description taxonomy-id base-media nitrogen-source  \\\n",
       "0                 WT      511145         M9        NH4Cl(1)   \n",
       "1                 WT      511145         M9        NH4Cl(1)   \n",
       "2                 WT      511145         M9        NH4Cl(1)   \n",
       "3                 WT      511145         M9        NH4Cl(1)   \n",
       "4                 WT      511145         M9        NH4Cl(1)   \n",
       "\n",
       "       phosphorous-source sulfur-source calcium-source clean genes  \n",
       "0  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)      {nagA}  \n",
       "1  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)      {clsA}  \n",
       "2  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)       {rph}  \n",
       "3  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)      {rpoC}  \n",
       "4  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)       {hfq}  \n",
       "\n",
       "[5 rows x 49 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os, sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from aledbmutil.mut import get_clean_mut_gene_list, is_genetic_mut\n",
    "\n",
    "mut_df[\"genetic\"] = mut_df[\"Details\"].apply(is_genetic_mut)\n",
    "mut_df[\"clean genes\"] = mut_df.apply(\n",
    "    lambda m: set(get_clean_mut_gene_list(m[\"Gene\"]) if m[\"genetic\"] else set()),  # For when using gene names instead of bnums\n",
    "    axis=1)\n",
    "mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _get_gene_bnum(gene_name):\n",
    "    gene_bnum = \"\"\n",
    "    gene_df = regdb_gene_df[regdb_gene_df[\"GENE_NAME\"]==gene_name]\n",
    "    if len(gene_df) > 0:\n",
    "        regdb_id = gene_df.iloc[0][\"GENE_ID\"]\n",
    "        g_synonym_df = gene_synonym_df[gene_synonym_df[\"OBJECT_ID\"]==regdb_id]\n",
    "        g_bnum_df = g_synonym_df[g_synonym_df[\"OBJECT_SYNONYM_NAME\"].str.contains('^b\\d{4}')]\n",
    "        if len(g_bnum_df) > 0:\n",
    "            gene_bnum = g_bnum_df.iloc[0][\"OBJECT_SYNONYM_NAME\"]\n",
    "    return gene_bnum\n",
    "\n",
    "assert(_get_gene_bnum(\"rpoB\")==\"b3987\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 3921/3921 [00:08<00:00, 465.56it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>...</th>\n",
       "      <th>supplement</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>base-media</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>calcium-source</th>\n",
       "      <th>clean genes</th>\n",
       "      <th>bnums</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>...</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>{nagA}</td>\n",
       "      <td>{b0677}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>{clsA}</td>\n",
       "      <td>{b1249}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3815859</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>...</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>{rph}</td>\n",
       "      <td>{b3643}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4187550</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>...</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>{rpoC}</td>\n",
       "      <td>{b3988}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4400313</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>{hfq}</td>\n",
       "      <td>{b4172}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  index  exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "0     9  42C    1    124        1         1       1.0    702352           DEL   \n",
       "1    10  42C    1    124        1         1       1.0   1308318           SNP   \n",
       "2    11  42C    1    124        1         1       1.0   3815859           DEL   \n",
       "3    12  42C    1    124        1         1       1.0   4187550           SNP   \n",
       "4    13  42C    1    124        1         1       1.0   4400313           SNP   \n",
       "\n",
       "  Sequence Change  ...                   supplement strain-description  \\\n",
       "0          Δ21 bp  ...  NaCl(0.5g/L) trace elements                 WT   \n",
       "1             G→C  ...  NaCl(0.5g/L) trace elements                 WT   \n",
       "2          Δ82 bp  ...  NaCl(0.5g/L) trace elements                 WT   \n",
       "3             C→T  ...  NaCl(0.5g/L) trace elements                 WT   \n",
       "4             A→C  ...  NaCl(0.5g/L) trace elements                 WT   \n",
       "\n",
       "  taxonomy-id base-media nitrogen-source      phosphorous-source  \\\n",
       "0      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "2      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "3      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "4      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "\n",
       "   sulfur-source calcium-source clean genes    bnums  \n",
       "0    MgSO4(0.24)     CaCl2(0.1)      {nagA}  {b0677}  \n",
       "1    MgSO4(0.24)     CaCl2(0.1)      {clsA}  {b1249}  \n",
       "2    MgSO4(0.24)     CaCl2(0.1)       {rph}  {b3643}  \n",
       "3    MgSO4(0.24)     CaCl2(0.1)      {rpoC}  {b3988}  \n",
       "4    MgSO4(0.24)     CaCl2(0.1)       {hfq}  {b4172}  \n",
       "\n",
       "[5 rows x 50 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "tqdm.pandas()\n",
    "\n",
    "# from pandarallel import pandarallel\n",
    "# pandarallel.initialize(progress_bar=True)\n",
    "\n",
    "# For whatever reason, parallel_apply's progress bars freeze,\n",
    "# which makes me think that parallel_apply crashes, though not sure.\n",
    "# tqdm's progress_apply successfully returns progress though.\n",
    "mut_df[\"bnums\"] = mut_df.progress_apply(\n",
    "# aledb_MG1655_ep_max_freq_mut_df[\"bnums\"] = aledb_MG1655_ep_max_freq_mut_df.parallel_apply(\n",
    "    lambda m: {_get_gene_bnum(gn) for gn in m['clean genes']},\n",
    "    axis=1)\n",
    "mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# key genes from Patrick's Ind Chem Tol GDV study that are commonly truncated.\n",
    "genes_of_interest= {\n",
    "    'b3783',  # rho\n",
    "    'b3938',  # metJ\n",
    "    'b3295',  # rpoA\n",
    "    'b3251',  # mreB\n",
    "    'b3650',  # spoT\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'rho': 16, 'mreB': 8, 'rpoA': 46, 'metJ': 10, 'spoT': 18})"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "\n",
    "\n",
    "mut_df[\"genes of interest\"] = mut_df.apply(\n",
    "    lambda m: m[\"bnums\"] & genes_of_interest,\n",
    "    axis=1)\n",
    "mut_genes_of_interest_df = mut_df[mut_df[\"genes of interest\"]!=set()].copy()\n",
    "display(Counter(mut_genes_of_interest_df.Gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'rho': 15, 'mreB': 8, 'rpoA': 40, 'metJ': 9, 'spoT': 18})"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Only want SNPs since we specifically AA substitutions.\n",
    "snp_genes_of_interest_df = mut_genes_of_interest_df[mut_genes_of_interest_df[\"Mutation Type\"]==\"SNP\"].copy()\n",
    "display(Counter(snp_genes_of_interest_df.Gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>substitution</th>\n",
       "      <th>variant source</th>\n",
       "      <th>conditions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>rho</td>\n",
       "      <td>G61E</td>\n",
       "      <td>ALEdb</td>\n",
       "      <td>{'temperature': '42 celsius', 'carbon-source':...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>mreB</td>\n",
       "      <td>E346A</td>\n",
       "      <td>ALEdb</td>\n",
       "      <td>{'temperature': '37 celsius', 'carbon-source':...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636</th>\n",
       "      <td>rho</td>\n",
       "      <td>K417I</td>\n",
       "      <td>ALEdb</td>\n",
       "      <td>{'temperature': '37 celsius', 'carbon-source':...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>653</th>\n",
       "      <td>rho</td>\n",
       "      <td>I382F</td>\n",
       "      <td>ALEdb</td>\n",
       "      <td>{'temperature': '37 celsius', 'carbon-source':...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1160</th>\n",
       "      <td>rpoA</td>\n",
       "      <td>T285I</td>\n",
       "      <td>ALEdb</td>\n",
       "      <td>{'temperature': '37 celsius', 'carbon-source':...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene substitution variant source  \\\n",
       "12     rho         G61E          ALEdb   \n",
       "176   mreB        E346A          ALEdb   \n",
       "636    rho        K417I          ALEdb   \n",
       "653    rho        I382F          ALEdb   \n",
       "1160  rpoA        T285I          ALEdb   \n",
       "\n",
       "                                             conditions  \n",
       "12    {'temperature': '42 celsius', 'carbon-source':...  \n",
       "176   {'temperature': '37 celsius', 'carbon-source':...  \n",
       "636   {'temperature': '37 celsius', 'carbon-source':...  \n",
       "653   {'temperature': '37 celsius', 'carbon-source':...  \n",
       "1160  {'temperature': '37 celsius', 'carbon-source':...  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# snp_genes_of_interest_df.to_csv('./data/aledb_snp_df.pkl')\n",
    "# out_df = pd.DataFrame(columns=['gene', 'AA sequence', 'substitution', 'variant source', 'conditions'])\n",
    "snp_genes_of_interest_df.rename(columns={\"Gene\": \"gene\"}, inplace=True)\n",
    "snp_genes_of_interest_df['substitution'] = snp_genes_of_interest_df.Details.apply(lambda s: s.split(' ')[0])\n",
    "snp_genes_of_interest_df['variant source'] = 'ALEdb'\n",
    "\n",
    "conditions = [\n",
    "    'temperature', 'carbon-source', 'supplement',\n",
    "    'strain-description', 'base-media', 'nitrogen-source',\n",
    "    'phosphorous-source', 'sulfur-source', 'calcium-source']\n",
    "snp_genes_of_interest_df['conditions'] = snp_genes_of_interest_df.apply(\n",
    "    lambda r: {c: r[c] for c in conditions},\n",
    "    axis=1)\n",
    "snp_genes_of_interest_df = snp_genes_of_interest_df[['gene', 'substitution', 'variant source', 'conditions']]\n",
    "snp_genes_of_interest_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_genes_of_interest_df.to_csv('./aledb_snp_df.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
